docs/proposals/component-registration.md

```diff
@@ -0,0 +1,378 @@
+  - [Use Cases](#use-cases)
+    - [Deployment](#deployment)
+    - [Debugging](#debugging)
+    - [Notification](#notification)
+    - [Reporting](#reporting)
+    - [Self-Healing](#self-healing)
+  - [Design](#design)
+    - [API Changes](#api-changes)
+      - [Component API](#component-api)
+      - [ComponentStatuses API](#componentstatuses-api)
+    - [Kubectl (CLI)](#kubectl-cli)
+    - [Storage](#storage)
+    - [Status Updates](#status-updates)
+  - [History](#history)
+  - [Example](#example)
+  - [Next Steps](#next-steps)
+
+<!-- END MUNGE: GENERATED_TOC -->
+
+## Abstract
+
+Kubernetes was designed to be extensible from the start, but one factor currently limits that capability: the hardcoding of what defines a Kubernetes cluster. To reach enterprise-class capabilities (high availability, notifications, self-healing, reporting, and maintainability) Kubernetes first needs to support dynamic component registration and readiness probes.
+
+## Terminology
+
+To clarify, a "component" in this context is a core part of the Kubernetes cluster, required or optional, that knows about Kubernetes and exposes a method for probing the liveness and readiness of each instance.
+
+The set of available/deployed/registered component types and/or instances may change over time.
+
+Currently there are three types of components in a vanilla Kubernetes deployment, each with a single instance: apiserver, controller-manager, and scheduler.
+
+Etcd should not be considered a component, because it is a dependency of the apiserver, does not know about Kubernetes, and thus cannot register itself or be expected to have both liveness and readiness probe endpoints. However, the current `/componentstatuses` endpoint includes etcd, and must continue to do so for reverse compatibility.
+
+
+## Use Cases
+
+This proposal primarily covers features that can be achieved internally by Kubernetes itself, but it's clear that there are also some desirable features that must be provided by external deployment, management, or monitoring systems. These external systems, however, require some features within Kubernetes to enable them. So while not all of these use cases can be solved by Kubernetes alone, we do want to make them possible.
+
+### Deployment
+
+As a k8s operator, I want a to be able to add/remove new components (instances or types) to a new or running cluster (ideally without manually registering it with the apiserver).
+
+As a k8s operator, I want a kubectl command that allows me to validate that the cluster has been deployed and is fully functional.
+
+As a k8s operator, I want a kubectl command that allows me to check which components are deployed/installed on a specific cluster in order to infer which features it supports.
+
+Note: Inferring features from components isn't very useful with vanilla k8s, but becomes more useful as 3rd party or optional components are added (e.g. OpenShift) and/or if components-controller is split into multiple components.
```

Comment:
Why aren't we inferring features from services?  These cases sound more like "yum install", but we've already got a "yum install".  I think having one object that represents the sum total of a set of items (the "component") seems interesting, but also has flaws in terms of our goal to build bottom up compositional patterns.  If you created something and it had a service that said it installed "dns", isn't that sufficient to know that DNS is installed?